mantidproject/mantid
# Crystal Field GOFit fitting

**Description of work.**
This PR adds the ability to perform a fit from the Crystal Field API which uses the GOFit python package. There are three GOFit
algorithms you can use: `regularisation`, `multistart` and `alternating`. The `alternating` algorithm is said to give very good results, even starting from initial parameters which are less than optimal.

I decided it would be neat to pass the gofit algorithm callable into the API so that we do not have a gofit import statement inside of mantid. I think this is a good idea because gofit will not be packaged with the mantid conda environment, so it makes sense for a user to import it and pass it in.

**To test:**
1. Make sure you have gofit pip installed in your environment `pip install gofit`
1. Open a script editor in mantid
2. Run the script below, and wait until it finishes.
```py
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
from CrystalField import CrystalField, CrystalFieldFit
import os
import gofit

minimizers = ['Nelder-Mead', 'BFGS']
iterations = [5, 10]

data_ws1 = Load(os.path.join(os.path.dirname(__file__), 'NdOs2Al10_5K35meV_Ecut_0to3ang_bp15V1.xye'))

np.random.RandomState(np.random.MT19937(np.random.SeedSequence(int(np.random.rand()*1e6))))

refpars = {'B20':0.19, 'B22':0.11, 'B40':-0.0004, 'B42':-0.002, 'B44':-0.012, 'B60':5.e-05, 'B62':0.00054, 'B64':-0.0006, 'B66':0.0008}
init_pars = {pn:refpars[pn]*((np.random.rand()-0.5)/2+1) for pn in refpars.keys()}

parameter_bounds = {'B20': (-0.3013,0.3013), 'B22': (-0.5219,0.5219), 'B40': (-0.004624,0.004624), 'B42': (-0.02068,0.02068), 'B44': (-0.02736,0.02736), 'B60': (-0.0001604,0.0001604), 'B62': (-0.001162,0.001162), 'B64': (-0.001273,0.001273), 'B66': (-0.001724,0.001724), 'IntensityScaling': (0.,10.), 'f0.FWHM': (0.1,5.0), 'f1.FWHM': (0.1,5.0), 'f2.FWHM': (0.1,5.0), 'f3.FWHM': (0.1,5.0), 'f4.FWHM': (0.1,7.0)}

cf = CrystalField('Nd', 'C2v', Temperature=5, FWHM=1, **init_pars)

fit2 = CrystalFieldFit(Model=cf, InputWorkspace=data_ws1, MaxIterations=50, Output='fit_gofit')
EvaluateFunction(Function=str(fit2.model.function), InputWorkspace=data_ws1, OutputWorkspace="guess")

fit2.gofit(algorithm_callable=gofit.regularisation, jacobian=True, maxit=3000)
#fit2.gofit(algorithm_callable=gofit.multistart,  jacobian=True, parameter_bounds=parameter_bounds, samples=100, maxit=3000, scaling=True)
#fit2.gofit(algorithm_callable=gofit.alternating,, parameter_bounds=parameter_bounds, samples=100, maxit=3000)

# Plots the data and print fitted parameters
l=plotSpectrum('fit_gofit_regularisation',[0,1,2],error_bars=False)
```
3. See that you get the following plot. Note that you might get a result which is not as good as this when running the `regularisation` and `multistart` algorithms, because the initial parameters are being randomized slightly.
![image](https://user-images.githubusercontent.com/40830825/179258397-a3f1f571-55ca-4010-b354-d110be99bf44.png)

5. Repeat this, but for the `multistart` and `alternating` algorithms

6. Now try testing the `alternating` algorithm with more difficult starting parameters:

```py
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
from CrystalField import CrystalField, CrystalFieldFit
import os
import gofit

#fully_random = True     # Use fully random starting pars
fully_random = False    # Just slightly randomize good pars
#fully_random = "MC"     # Use the Monte Carlo parameter estimation routine to refine fully random initial pars.
do_plot = True

minimizers = ['Nelder-Mead', 'BFGS']
iterations = [5, 10]

dir = os.path.dirname(__file__)
data_ws1 = Load(os.path.join(dir, 'NdOs2Al10_5K35meV_Ecut_0to3ang_bp15V1.xye'))

np.random.RandomState(np.random.MT19937(np.random.SeedSequence(int(np.random.rand()*1e6))))

if fully_random:
    parnames = ['B20', 'B22', 'B40', 'B42', 'B44', 'B60', 'B62', 'B64', 'B66']
    init_pars = {pn:(np.random.rand()-0.5)*2 for pn in parnames}
    
    if isinstance(fully_random, str) and fully_random.lower() == 'mc':
    # Use the "Monte Carlo" estimation routine to refine the random parameters a bit
        cf = CrystalField('Nd', 'C2v', Temperature=5, FWHM=1, **init_pars)
        cf.PeakShape = 'Lorentzian'
        cf.IntensityScaling = 2   # Scale factor if data is not in absolute units (mbarn/sr/f.u./meV), will be fitted.
        fit = CrystalFieldFit(Model=cf, InputWorkspace=data_ws1,MaxIterations=200)
        fit.estimate_parameters(EnergySplitting=20,
                                Parameters=parnames,
                                NSamples=1000)
        fit.select_estimated_parameters(1)
        init_pars = {pn:fit.model[pn] for pn in parnames}
else:
    # Just slightly randomize reference parameters
    parnames = ['B20', 'B22', 'B40', 'B42', 'B44', 'B60', 'B62', 'B64', 'B66']
    refpars = {'B20':0.19, 'B22':0.11, 'B40':-0.0004, 'B42':-0.002, 'B44':-0.012, 'B60':5.e-05, 'B62':0.00054, 'B64':-0.0006, 'B66':0.0008}
    init_pars = {pn:refpars[pn]*((np.random.rand()-0.5)/2+1) for pn in parnames}

cf = CrystalField('Nd', 'C2v', Temperature=5, FWHM=1, **init_pars)

fit = CrystalFieldFit(Model=cf, InputWorkspace=data_ws1, MaxIterations=50)
fit.gofit(algorithm_callable=gofit.alternating,, parameter_bounds=parameter_bounds, samples=100, maxit=3000)
```

Fixes #33864

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
